crates/jstz_api/src/console.rs

```diff
@@ -23,97 +21,66 @@ use boa_engine::{
     Context, JsArgs, JsNativeError, JsResult, JsValue, NativeFunction,
 };
 use boa_gc::{empty_trace, Finalize, GcRefMut, Trace};
-use jstz_core::{host::HostRuntime, runtime, value::IntoJs};
-use jstz_crypto::{hash::Blake2b, public_key_hash::PublicKeyHash};
-use serde::{Deserialize, Serialize};
-
-use tezos_smart_rollup::prelude::debug_msg;
-
-pub const LOG_PREFIX: &str = "[JSTZ:SMART_FUNCTION:LOG] ";
-
-#[derive(Serialize, Deserialize)]
-pub struct LogRecord {
-    pub contract_address: PublicKeyHash,
-    pub request_id: String,
-    pub level: String,
-    pub text: String,
-}
-
-impl Display for LogRecord {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        f.write_str(&serde_json::to_string(self).map_err(|_| std::fmt::Error)?)
-    }
-}
-
-impl FromStr for LogRecord {
-    type Err = serde_json::Error;
-    fn from_str(json: &str) -> Result<Self, Self::Err> {
-        serde_json::from_str(json)
+use jstz_core::value::IntoJs;
+use log::js_log;
+pub use log::{set_console_logger, ConsoleLog, LogData, LogLevel};
+
+mod log {
+    use boa_engine::{Context, JsNativeError, JsResult};
+    use serde::{Deserialize, Serialize};
+    use std::cell::Cell;
+
+    #[derive(Serialize, Deserialize)]
+    pub enum LogLevel {
+        ERROR,
+        WARN,
+        INFO,
+        LOG,
     }
-}
-
-/// This represents the different types of log messages.
-#[derive(Debug)]
-enum LogMessage {
-    Log(String),
-    Info(String),
-    Warn(String),
-    Error(String),
-}
 
-impl LogMessage {
-    fn log(self, rt: &impl HostRuntime, console: &Console) {
-        match console {
-            Console::Proto {
-                groups,
-                contract_address,
-                operation_hash,
-            } => {
-                let indent = 2 * groups.len();
-                let log_record = LogRecord {
-                    contract_address: contract_address.clone(),
-                    request_id: operation_hash.to_string(),
-                    level: self.level().to_string(),
-                    text: " ".repeat(indent) + self.message(),
-                }
-                .to_string();
-                rt.write_debug(&(LOG_PREFIX.to_string() + &log_record + "\n"));
-            }
-            Console::Cli { groups } => {
-                let indent = 2 * groups.len();
-                let symbol = self.symbol();
-                for line in self.message().lines() {
-                    debug_msg!(rt, "[{symbol}] {:>indent$}{line}\n", "");
-                }
+    impl LogLevel {
+        pub fn symbol(&self) -> char {
+            match self {
+                LogLevel::ERROR => '🔴',
+                LogLevel::WARN => '🟠',
+                LogLevel::INFO => '🟢',
+                LogLevel::LOG => '🪵',
             }
         }
     }
+    #[derive(Serialize, Deserialize)]
+    pub struct LogData {
```

Comment:
I would've suggested `LogRecord`, but I assume this was chosen to avoid the name collision with `LogRecord` in `jstz_proto`

Reply:
yes, i'm open to renaming it if you have better suggestion...